src/exercises/Exercises.js

```diff
@@ -71,7 +81,10 @@ export default function Exercises({ api, articleID }) {
     }
 
     setCurrentIndex(newIndex);
-    setCurretBookmarkToStudy(bookmarksToStudyList[newIndex]);
+    setCurrentBookmarkToStudy(bookmarksToStudyList[newIndex]);
+    api.wordsSimilarTo(bookmarksToStudyList[newIndex].id, (words) => {
```

Comment:
there's quite a bit of duplication between these two branches of the conditional. 

besides that, could we move the wordsSimilarTo to the useEffect of the MultipleChoice.js? it's only used there, so the principle encapsulation would recommend that we're keeping it there and don't let Exercises.js worry about it. 

Do you see what I mean?

Reply:
If you move the `wordsSimilartTo` and you remove duplication, you should be able to write something like: 

```
  function initializeExercises(bookmarks, title) {
    setbookmarksToStudyList(bookmarks);
    NUMBER_OF_EXERCISES = bookmarks.length;
    setCurrentBookmarkToStudy(bookmarks[currentIndex]);
    setTitle(title);
  }

  if (!bookmarksToStudyList) {
    if (articleID) {
      api.bookmarksForArticle(articleID, (bookmarks) => {
        api.getArticleInfo(articleID, (data) => {
          setArticleInfo(data);
          initializeExercises(bookmarks, 'Exercises for "' + data.title + '"');
        });
      });
    } else {
      api.getUserBookmarksToStudy(NUMBER_OF_EXERCISES, (bookmarks) => {
        initializeExercises(bookmarks, "Exercises");
      });
    }
  }
```

Reply:
Fully agree. I've moved the whole `wordsSimilarTo` into `MultipleChoice` in the 0eb2566ff2ec9ea487b64700ce88edd4a2d4c859 commit. It works but now one has to wait a few seconds before the buttons are rendered as the api call needs to happen first. So I've added in the loading animation as a buffer. If this is ok with you, it's fine by me too! 
By having the code in `Exercises` I could make sure the exercises were only rendered once all api calls had successfully been completed, avoiding this slower render for every multiple choice exercise.

Reply:
Thanks!

I've tried to make the /similar_words faster but I still get the 
chance to see the half-rendered exercise.

![image](https://user-images.githubusercontent.com/464519/116275382-3fa57480-a78c-11eb-9c23-8c25de518bf9.png)

We have to find a way of improving this. 
Maybe open an issue and let's move on for now?

Reply:
Option could be rendering already the text, and adding the loading only in the place of the buttons. Like that the exercise jumps around  much less!

Reply:
Also fixed the `useEffect` to only be called upon render - this should fix some issues and the buttons appear quite fast now.